File: src/系统/03．技能系统/00．技能模板+函数/02．通用函数/01．便捷短函数集合/01．不死函数.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
--- 便捷短函数 - 不死、等一次性效果
local jass = require("jass.common")
local ____require_result_0 = require("系统.04．伤害系统.00．伤害计算.06．伤害修正回调")
local registerDamageModifier = ____require_result_0.registerDamageModifier
local GetHandleId = jass.GetHandleId
local GetUnitState = jass.GetUnitState
local _____4E0D_6B7B_96C6_5408 = {}
local function ____on_4E0D_6B7B_4F24_5BB3_4FEE_6B63(context)
    local target = context.target
    if target == nil or target == 0 then
        return context.currentDamage
    end
    local _____5355_4F4DID = GetHandleId(target)
    if _____5355_4F4DID == 0 then
        return context.currentDamage
    end
    if _____4E0D_6B7B_96C6_5408[_____5355_4F4DID] == nil then
        return context.currentDamage
    end
    local _____5F53_524D_8840_91CF = GetUnitState(target, jass.UNIT_STATE_LIFE)
    if _____5F53_524D_8840_91CF - context.currentDamage < 1 then
        return _____5F53_524D_8840_91CF - 1
    end
    return context.currentDamage
end
local _____4FEE_6B63_5668_5DF2_6CE8_518C = false
local function _____786E_4FDD_4FEE_6B63_5668_6CE8_518C()
    if _____4FEE_6B63_5668_5DF2_6CE8_518C then
        return
    end
    _____4FEE_6B63_5668_5DF2_6CE8_518C = true
    registerDamageModifier(____on_4E0D_6B7B_4F24_5BB3_4FEE_6B63, 60)
end
--- 令单位进入不死状态，受到致命伤害时保留1点血量。
-- 需调用方自行管理生命周期（搭配Buff到期回调等调用 移除单位不死）。
____exports["令单位不死"] = function(_____5355_4F4D)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return
    end
    _____786E_4FDD_4FEE_6B63_5668_6CE8_518C()
    local _____5355_4F4DID = GetHandleId(_____5355_4F4D)
    if _____5355_4F4DID == 0 then
        return
    end
    _____4E0D_6B7B_96C6_5408[_____5355_4F4DID] = true
end
--- 移除单位的不死状态。
____exports["移除单位不死"] = function(_____5355_4F4D)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return
    end
    local _____5355_4F4DID = GetHandleId(_____5355_4F4D)
    if _____5355_4F4DID == 0 then
        return
    end
    __TS__Delete(_____4E0D_6B7B_96C6_5408, _____5355_4F4DID)
end
--- 查询单位是否处于不死状态。
____exports["单位是否不死"] = function(_____5355_4F4D)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return false
    end
    local _____5355_4F4DID = GetHandleId(_____5355_4F4D)
    return _____5355_4F4DID ~= 0 and _____4E0D_6B7B_96C6_5408[_____5355_4F4DID] ~= nil
end
return ____exports

File: src/系统/03．技能系统/00．技能模板+函数/02．通用函数/01．便捷短函数集合/01．不死函数.js
/** @noSelfInFile */
/**
 * 便捷短函数 - 不死、等一次性效果
 */
const jass = require("jass.common");
const { registerDamageModifier } = require("系统.04．伤害系统.00．伤害计算.06．伤害修正回调");
const GetHandleId = jass.GetHandleId;
const GetUnitState = jass.GetUnitState;
const 不死集合 = {};
function on不死伤害修正(context) {
    const target = context.target;
    if (target == null || target === 0)
        return context.currentDamage;
    const 单位ID = GetHandleId(target);
    if (单位ID === 0)
        return context.currentDamage;
    if (不死集合[单位ID] == null)
        return context.currentDamage;
    const 当前血量 = GetUnitState(target, jass.UNIT_STATE_LIFE);
    if (当前血量 - context.currentDamage < 1) {
        return 当前血量 - 1;
    }
    return context.currentDamage;
}
let 修正器已注册 = false;
function 确保修正器注册() {
    if (修正器已注册)
        return;
    修正器已注册 = true;
    registerDamageModifier(on不死伤害修正, 60);
}
/**
 * 令单位进入不死状态，受到致命伤害时保留1点血量。
 * 需调用方自行管理生命周期（搭配Buff到期回调等调用 移除单位不死）。
 */
export function 令单位不死(单位) {
    if (单位 == null || 单位 === 0)
        return;
    确保修正器注册();
    const 单位ID = GetHandleId(单位);
    if (单位ID === 0)
        return;
    不死集合[单位ID] = true;
}
/**
 * 移除单位的不死状态。
 */
export function 移除单位不死(单位) {
    if (单位 == null || 单位 === 0)
        return;
    const 单位ID = GetHandleId(单位);
    if (单位ID === 0)
        return;
    delete 不死集合[单位ID];
}
/**
 * 查询单位是否处于不死状态。
 */
export function 单位是否不死(单位) {
    if (单位 == null || 单位 === 0)
        return false;
    const 单位ID = GetHandleId(单位);
    return 单位ID !== 0 && 不死集合[单位ID] != null;
}
